### ISSUE TITLE
Y.Type is always AbstractType after `applyUpdate`

### ISSUE THREAD

**Issue by himself65** [yjs/yjs#563]
```ts
const doc = Y.Doc()
const map = doc.getMap('1')

const doc2 = Y.Doc()
applyUpdate(doc2, encodeStateAsUpdate(doc))
{
  const map = doc.get('1')
  // not ok
  ok((map instanceof Y.Map) === true)
}
```

**Comment by himself65** [comment]
I'm a little confuse because this feature(I think it's not a bug because you mentioned it before) could cause our event listener to treat the type incorrectly

**Comment by dmonad** [comment]
The top-level types are not defined globally - every client must define them for themselves. The idea is that one client could have another "view" on the data than another client. E.g. it is possible to do `ydoc.get('document', Y.XmlFragment)` (read the document as a XML data type), and another client could do `ydoc.get('document', Y.Array)` (read the document as an Array data structure). In the future, there might be other data types that you can use to interpret the shared data. However, I don't really recommend doing this in your case..

Another reason for not sharing type definitions globally is that it's impractical to share the type definitions of the root data types. In case the definitions mismatch (user1: `ydoc.get('document', Y.Array)`, but user2: `ydoc.get('document', Y.XmlFragment)`) would have to throw an exception if we enforce everyone to use the same types. Hence, we can as well allow it, as it is technically possible.

TLDR: My recommendation is to always specify what you are expecting. Don't do `ydoc.get('doc')`, do `ydoc.get('doc', Y.Array)` instead. Your application should know what it expects.
